pyro/ops/einsum/contract.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import, division, print_function
+
+import importlib
+from collections import Counter
+
+import torch
+from opt_einsum.sharing import einsum_cache_wrap
+
+from pyro.ops.einsum.paths import optimize
+
+
+@einsum_cache_wrap
+def _einsum(equation, *operands, **kwargs):
+    backend = kwargs.pop('backend', 'numpy')
+    if backend == 'torch':
+        # provide np.einsum interface for torch.einsum.
+        return torch.einsum(equation, operands)
+    einsum = getattr(importlib.import_module(backend), 'einsum')
+    return einsum(equation, *operands, **kwargs)
+
+
+class ContractExpression(object):
```

Comment:
This is the entirety of the new logic, similar to opt_einsum's [ContractExpression](https://github.com/dgasmith/opt_einsum/blob/dbf284b/opt_einsum/contract.py#L575).

Reply:
It looks like the main piece of functionality in `opt_einsum` left out of this version is constant terms.  Were we ever using those?

Reply:
We were not using constant terms. I don't foresee us using those. Hopefully when `ubersum` settles down we can try to move this back upstream.